.github/workflows/python-app.yml

```
@@ -0,0 +1,36 @@
+# This workflow will install Python dependencies, run tests and lint with a single version of Python
+# For more information see: https://help.github.com/actions/language-and-framework-guides/using-python-with-github-actions
+
+name: Tests
+
+on:
+  push:
+    branches: [ main ]
+  pull_request:
+    branches: [ main ]
+
+jobs:
+  ubuntu_2004:
+
+    runs-on: ubuntu-20.04
+
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up Python 3.8
+        uses: actions/setup-python@v2
+        with:
+          python-version: 3.8
+      - name: Install dependencies
+        run: |
+          python -m pip install pip==20.3.1
+          pip install flake8==3.8.4 pytest==6.1.2
```

Assuming that the project are using semantic versioning, we should be able to remove the bugfix/patch version as we will know that the API will not change between X.Y.1 and X.Y.3

```suggestion
          pip install flake8==3.8 pytest==6.1
```

At the same time if projects are serious about semver, we could also just pin the major verrsion, as in theory minor version only add features, without removing them ... but this is not always true in practice.


```suggestion
          pip install flake8==3 pytest==6
```

Also, projects are starting to use calendar version ...to make it clear that there is no major or minor version :) ...but we can still pin only on `pip==20.3` as any pip 20.3.1 or 20.3.4 should behave the same.

---

Indeed. But if we want completely deterministic dependencies, we also need to pin the dependencies' dependencies, such as with `pip freeze`, or with a `Pipfile.lock` using [Pipenv](https://pipenv.pypa.io/en/latest/).

But I suppose Paver can also offer that.

---

paver is just a simple library that implements something like `make` in Python.

It's true that in theory we should also pin all the deps. I remember seeing this in NodeJs projects.

In practice, I don't remember test failures generated by secondary libraries... and if we found such a failure, we can pin just that library.

I am not a big fan on pinning everything... there should be some slack :)

But we can give pypenv a try.